---
title: "iSEE User's Guide"
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: aaron.lun@cruk.cam.ac.uk
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevin.rue-albrecht@kennedy.ox.ac.uk
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
abstract: >
  Abstract for iSEE
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{iSEE User's Guide}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---
<!-- as biocviews: -->
<!-- Visualization, GUI, DimensionReduction,  -->
<!--         FeatureExtraction, Clustering, Transcription,  -->
<!--         GeneExpression, Transcriptomics, SingleCell -->

<!-- **Package**: `r Rpackage("iSEE")` -->

<!-- **Authors**: `r packageDescription("iSEE")[["Author"]]` -->

<!-- **Version**: `r packageDescription("iSEE")$Version` -->

**Compiled date**: `r Sys.Date()`

**Last edited**: 2017-12-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Getting started

<!-- logo? MOS DEF :) -->

`r Biocpkg("iSEE")` is an R package distributed as part of the
[Bioconductor](http://bioconductor.org) project. To install the package,
start R and enter:

```{r installation, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
```

To load and attach the package to your current workspace, enter:

```{r library, message=FALSE, warning=FALSE}
library("iSEE")
```


# Introduction

[]what does iSEE do []

## Citation info

If you use `r Biocpkg("iSEE")` for your analysis, please cite it as here below:

```{r citation}
citation("iSEE")
```



# Quick start

[]want to dive deep right away? here's what you need to know[]

i.e. get your SCE object
load lib and launch app
follow the tour

.. this one that follows is then more detailed, starting from...



# Using the application

param descriptions
ways to change default settings?


if applicable:
- locally
- public instance to demo
- deploying to your shiny server


# Getting to know the user interface and the functionality

header, sidebar, main panel

## Specifical subsection on the help/tour?


# Functionality

aka description of plot types

concept of brushing and linking - here?

this section has for each type of plot a screenshot of plot + plot settings


# Bug reporting/Devel Version


The GitHub repository for `r Biocpkg("iSEE")` is https://github.com/csoneson/iSEE, where you can find the development version of the package. This is also the place to file an issue, report a bug, or provide a pull request.


# Use cases

## Allen dataset





## TCGA dataset


-> link to youtube screencast where we do it?!



## Using the `ExperimentHub`

```{r ExperimentHub, eval=FALSE}
stopifnot(
  require(ExperimentHub),
  require(SingleCellExperiment),
  require(irlba),
  require(Rtsne),
  requireNamespace("scater")
)
ehub <- ExperimentHub()
eh1 <- ehub[["EH1"]] # an ExpressionSet
se1 <- as(eh1, "SummarizedExperiment")
sce1 <- as(se1, "SingleCellExperiment")
sce1 <- scater::runPCA(sce1, exprs_values = "exprs")
irlba_out <- irlba(sce1@assays@.xData$data$exprs)
tsne_out <- Rtsne(irlba_out$v, pca = FALSE, perplexity = 50, verbose = TRUE)
reducedDim(sce1, "TSNE") <- tsne_out$Y
if (interactive()) { iSEE(sce1) }
```


## Using online file sharing systems

Active projects may require regular updates to the `SingleCellExperiment`
object, that must be iteratively re-distributed to collaborators.

One possible choice is to store the up-to-date `SingleCellExperiment` object
in an file produced by the `saveRDS` command, and to host this file in
an online file sharing system (*e.g.*, Drobpox).

Users may then download the file and launch the `iSEE` application as follows:

```{r curl_download, eval=FALSE}
library(curl)
rdsURL <- "https://my.file.sharing.website.com/URI"
temp_file <- tempfile(pattern = "iSEE_", fileext = ".rds")
message("Downloading URL to temporary location: ", temp_file)
curl_download(url = rdsURL, destfile = temp_file, quiet = FALSE)
sce <- readRDS(temp_file)
if (interactive()) { iSEE(sce) }
```

To save the downloaded file to a persistent location, users may adapt
the following code chunk:

```{r copy_curl, eval=FALSE}
newLocation <- "/path/of/your/choice/new_file_name.rds"
file.copy(from = temp_file, to = newLocation)
```



# Further development

Additional functionality for the `r Biocpkg("iSEE")` will be added in the future.

Improvements, suggestions, bugs, issues and feedback of any type can be sent to marinif@uni-mainz.de.


# FAQ

**Q: Can you implement a 'Copy to clipboard' button in the code editor?**

A: This is not necessary, as one can click anywhere in the code editor and
instantly select all the code using a keyboard shortcut that depends on
your operating system.

**Q: When brushing with a transparency effect,
it seems that data points in the receiving plot are not subsetted correctly.**

A: The subsetting is correct. What you see is an artefact of overplotting:
in areas excessively dense in points, transparency ceases to be an effective
visual effect.





# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```


# References {.unnumbered}
 -> if we use them







