---
title: >
  `r Biocpkg("iSEE")` User's Guide
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: aaron.lun@cruk.cam.ac.uk
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevin.rue-albrecht@kennedy.ox.ac.uk
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
abstract: >
  Abstract for iSEE
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{iSEE User's Guide}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: iSEE.bib
---
<!-- as biocviews: -->
<!-- Visualization, GUI, DimensionReduction,  -->
<!--         FeatureExtraction, Clustering, Transcription,  -->
<!--         GeneExpression, Transcriptomics, SingleCell -->

<!-- **Package**: `r Rpackage("iSEE")` -->

<!-- **Authors**: `r packageDescription("iSEE")[["Author"]]` -->

<!-- **Version**: `r packageDescription("iSEE")$Version` -->

**Compiled date**: `r Sys.Date()`

**Last edited**: 2017-12-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  error=FALSE,
  warning=FALSE,
  message=FALSE
)
```

# Introduction 

<!-- logo? MOS DEF :) -->

`r Biocpkg("iSEE")` is a [Bioconductor](http://bioconductor.org) package that provides an interactive Shiny-based graphical user interface for exploring data stored in `SummarizedExperiment` objects, including row- and column-level metadata. 
Particular attention is given to single-cell data in a `SingleCellExperiment` object with visualization of dimensionality reduction results, e.g., from principal components analysis (PCA) or _t_-distributed stochastic neighbour embedding (_t_-SNE, [@van2008visualizing]).

To install the package, start R and enter:

```{r installation, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
```

To load and attach the package to your current workspace, enter:

```{r library}
library("iSEE")
```

# Quick start {#quickstart}

Do you want to dive deep in using `r Biocpkg("iSEE")` right away? 
Here's what you need to do:

- Get a dataset to work with, such as the `allen` dataset from the `r Biocpkg("scRNAseq")` package.
Transform it into a `SingleCellExperiment` object, where you can store the reduced dimension representations (here, PCA and _t_-SNE):

```{r qs-data}
library(scRNAseq)
data(allen)
library(scater)
sce <- as(allen, "SingleCellExperiment")
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)
sce <- runPCA(sce)
sce <- runTSNE(sce)
sce
```

- Create the Shiny app by running the `iSEE` function. 
See `?iSEE` for how to tweak the initial settings of the interface.
By default, the function will generate one panel of every type - see below for details on the different panel types. 

```{r qs-create}
app <- iSEE(sce)
```

- Launch the app using the `runApp` function.

```{r qs-launch, eval=FALSE}
shiny::runApp(app)
```

- Once you have started the app, look in the upper right corner for a **question mark** icon, and click on the button for an introductory tour.
This will perform an interactive tour of the app, based on the `r CRANpkg("rintrojs")` package [@ganz2016rintrojs].
During this tour, you will be taken through the different components of `r Biocpkg("iSEE")` and learn the basic usage mechanisms by doing: 
the highlighted elements will be responding to the user's actions, while the rest of the UI will be shaded. 
You can move forward and backward along the tour by clicking on the `Next`/`Back` buttons, or also using the arrow keys. 
To exit the tour, either click on `Skip`, or simply click outside of the highlighted UI element.

- Once you are done generating the plots, click on the **wrench** icon in the upper right corner, and click on the button for exporting the R code. 
This will open a modal popup where the R code used to generate the plot is displayed in a `r CRANpkg("shinyAce")`-based text editor. 
Select parts/all of it to copy-and-paste it into your analysis script/Rmarkdown file.

<!--
# Using the application

The `iSEE` function has a number of parameters that can be tweaked to adjust the initial settings of the interface.

- `se` is a `SingleCellExperiment` or `SummarizedExperiment` object, the only mandatory parameter for launching `iSEE`.

- `redDimArgs` is a `DataFrame` similar to that produced by the `redDimPlotDefaults` function, specifying initial parameters for the reduced dimension plots. 
Correspondingly, `colDataArgs` (e.g. created via `colDataPlotDefaults`) specifies the initial parameters for the column data plots, and `geneExprArgs` (via `geneExprPlotDefaults`) for the gene expression plots. 
`geneStatArgs` (a `DataFrame` structured as the output of  `geneStatTableDefaults`) specifies the initial parameters for the gene statistics tables

- `redDimMax`, `colDataMax`, and `geneExprMax` are integer scalars specifying the maximum number of plots (respectively, reduced dimension, column data, and gene expression) in the main interface.
Similarly, `geneStatMax` is an integer scalar specifying the maximum number of gene statistic tables in the interface.

- `initialPanels` is a `DataFrame` specifying which panels should be created at the  initialization. 
This should contain a `Name` character field (specifying the identity of the panel, e.g., "Reduced dimension plot 1", "Gene statistics table 2") and optionally, a `Width` integer field (ranging from 2 to 12) and a `Height` integer field (ranging from 400 to 1000 pixels).
If not specified, it will generate one plot per type, together with one gene statistics table.

- `annot.orgdb` is an `org.*.eg.db` annotation object from which Entrez identifiers can be retrieved. 
If this is specified, gene information will be retrieved upon selection of particular genes in the data table.

- `annot.keytype` is a string specifying the keytype to use to query `annot.orgdb`, while `annot.keyfield` is a string specifying the field of `rowData(se)` containing the keys of type `annot.keytype`. 
If `NULL`, the row names of `se` are used as the keys.

- `colormap` is an `ExperimentColorMap` object that defines custom color maps to apply to individual `assays`, `colData`, and `rowData` covariates. 
For more details on how to create this object, please see `?ExperimentColorMap`.

Upon calling `iSEE`, a Shiny App is created for interactive data exploration of the `SingleCellExperiment`/`SummarizedExperiment` object.
This can be launched in your browser by calling `runApp`.

## Using `r Biocpkg("iSEE")` from a demo public instance

[]if we want to allow this[]

## Deploying `r Biocpkg("iSEE")` to your Shiny Server

[]same here[]
-->

# Description of the user interface 

## Header

The layout of the `r Biocpkg("iSEE")` user interface is based mostly on the the `r CRANpkg("shinydashboard")` package.
The dashboard header contains several dropdown menus, where a number of help elements are available:

- The "Click me for a quick tour" button activates the `r CRANpkg("rintrojs")` based tour. Some more detail on this is included in Section \@ref(quickstart)
- The "Open the vignette (web)" button opens in a new tab the vignette for `r Biocpkg("iSEE")` from the Bioconductor website (internet connection is required). Please keep in mind it will be referring to the most current release version, which may differ from the version installed on your system)
- The "Open the vignette (local)" button is a fallback solution that fetches the HTML vignette of `r Biocpkg("iSEE")` as it is available on the system
- The "Session information" button provides the output from `sessionInfo()`, which should be included in case of reporting a bug/issue.
- The "About iSEE" displays a modal popup containing information about the authors and how to cite `r Biocpkg("iSEE")`.

A very useful element is the "Extract the R code!" button marked with the **magic wand** icon, in the dropdown menu labelled with the **wrench** icon. 
By clicking on this, the user opens up a modal popup window, where an instance of a text editor (based on the `r CRANpkg("ShinyAce")` package) contains the code snippet required to reproduce the plots created in the live session. 
This is typically used at the end of each exploration to extract a reproducible snapshot of the commands, which can be further edited to finalize the plots (e.g., for publication). 

Another functionality in the `iSEE diagnostics` dropdown, a button clickable marked with the chain icon, allows you to obtain a graph representation of the existing links and brushes among your open plot and table panels (coded with the same colors as in the app). This can be very useful once you open a multitude of panels, so that you can easily get an overview of their structure.

## Sidebar

The dashboard sidebar of `r Biocpkg("iSEE")` contains buttons for creating new plots or tables (referred to as "panels") in the interface.
The maximum number of panels of each type are defined when launching the app. 
Upon clicking these buttons, additional panels of the corresponding type are inserted into the main body of the app.
Color-coded tabs in the sidebar contain buttons to change the panel order (**up** or **down** icons) or to remove panels entirely (**trash** icon).
The width and height of each panel can also be adjusted by clicking on the **gear** icon.

## Body

The main element in the body of `r Biocpkg("iSEE")` is the combination of panels, generated (and optionally linked to one another) according to the user's actions.
The number and identity of the panels and their inter-relationships can also be specified at initialization by passing appropriate arguments to `iSEE`.
The main explanation on how the different plots and tables work is presented in Section \@ref(functionality).

# Description of `iSEE` functionality {#functionality}

## Overview

There are currently four different panel types that can be generated with `iSEE`:

- Reduced dimension plots
- Column data plots
- Gene expression plots
- Gene statistics tables

For each plot panel, three different sets of parameters will be available in collapsible boxes:

- Plotting parameters, controlling parameters specific to each type of plot. 
- Coloring parameters, specifying which sample (meta)data will be used to color the points.
- Brushing parameters, controlling the brushing relationship to other plots. 

## Reduced dimension plots

If a `SingleCellExperiment` object is supplied to `iSEE`, any reduced dimension results (e.g., from PCA or _t_-SNE) will be extracted from the `reducedDim` slot.
These results will be used to construct a two-dimensional reduced dimension plot where each point is a sample, to facilitate efficient exploration of high-dimensional datasets. 
The plotting parameters control the `reducedDim` slot to be displayed, as well as the two dimensions to plot against each other. 
Note that `iSEE` does not calculate the reduced dimension representations, so they must be present in the provided object.

## Column data plots

A column data plot involves visualizing sample metadata stored in the `SummarizedExperiment` object.
Different fields can be used for the x- and y-axis by selecting appropriate values in the plotting parameters.
This plot can assume various forms, depending on the nature of the data on the x- and y-axes: 

- If the y-axis is continuous and the x-axis is categorical, violin plots are generated (grouped by the x-axis factor).
- If the y-axis is categorical and the x-axis is continuous, horizontal violin plots are generated (grouped by the y-axis factor).
- If both are continuous, a scatter plot is generated.
- If both are categorical, a plot of squares is generated where the area of each square is proportional to the number of samples with a particular combination of levels.

An x-axis setting of "None" is considered to be categorical.

## Gene expression plots

A gene expression plot visualizes the expression values for a particular gene across the samples on the y-axis.
This usually results in a (grouped) violin plot, if the x-axis is set to "None" or a categorical variable; 
or a scatter plot, if the x-axis is another continuous variable.
(That said, if there are categorical values for the expression values, these will be handled as described in the column data plots.)

Gene selection for the y-axis is achieved by using a _linked gene statistics table_ in another panel.
Clicking on a gene in the table will automatically change the gene expression values plotted on the y-axis.
Alternatively, the gene name can be directly entered as text, corresponding to an entry of `rownames(se)`.
Note that this will not be effective if `se` contains no row names.

The x-axis covariate can also be selected from the plotting parameters.
This can be "None", column data, or the expression values of another gene (also identified using a linked table or via text).
The measurement units are selected as one of the `assays(se)`, which will apply to both the x- and y-axes.

## Gene statistics tables

A gene statistics table contains the values of the `rowData` slot for the `SingleCellExperiment`/`SummarizedExperiment` object. 
If none are specified, a column named `Present` is added and set to `TRUE` for all available genes, to avoid crashing `DT::datatable`.
Typically, these tables are used to link to other plots to determine the genes to use for plotting (or coloring).
However, they can also retrieve gene-specific annotation on the fly via the `r CRANpkg("rentrez")` package, provided the `annot.orgdb`, `annot.keytype`, and `annot.keyfield` parameters are specified.

## Coloring plots by sample attributes 

Coloring of points (i.e., samples) on each plot can be achieved in different ways.

- The default is no color scheme ("None" in the radio button).
This results in black points.
- Any column of `colData(se)` can be used.
The plotting routine will automatically adjust the scale to use based on whether the chosen column is continuous or categorical.
- The expression value of a particular gene in each sample can be used.
The gene can be chosen either via a linked gene table or text input (as described for gene expression plots).
Users can also specify the `assays` from which expression values are extracted.

## Linking plots with brushing {#brushing}

To link one plot to another, users can instruct a plotting panel to receive a brush from another (transmitting) plot, using the appropriate field in the brushing parameters box.
Once this is done, brushing on the transmitting plot will affect the receiving plot in a variety of ways:

- If the brushing effect is set to "Restrict", only the subset of points selected in the transmitter will be plotted in the receiver.
- If set to "Color", the selected subset of points will be plotted in the receiver with a color that can be selected via the `r CRANpkg("colourpicker")` package. 
- If set to "Transparent", the selected subset will be drawn with no transparency, while all non-selected points will be plotted with the specified alpha value.

## Zooming in and out

This is possible by first brushing to select a region of interest.
Double-clicking on the brushed area will zoom into the selected area.
To zoom out to the original plot, double-click outside of any brushed area.

# Use cases

## Use case I: Basic exploration of the Allen dataset

### Loading the data 

In this section, we illustrate how `r Biocpkg("iSEE")` can be used to explore the `allen` single-cell RNA-seq data set from  the `r Biocpkg("scRNAseq")` package.
It contains expression values for 379 cells from the mouse visual cortex [@tasic2016adult]. 
We start by converting the provided `SummarizedExperiment` object to a `SingleCellExperiment` object and normalize the expression values. 

```{r allen-dataset}
library(scRNAseq)
data(allen)
class(allen)
library(scater)
sce <- as(allen, "SingleCellExperiment")
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)
```

Next, we apply Principal Components Analysis (PCA) and t-distributed Stochastic Neighbour Embedding (t-SNE) to generate two reduced dimension representations of the cells. 
Note that all calculations (dimension reduction, clustering, etc) must be performed *before* passing the object to `iSEE`.

```{r allen-dataset-2}
sce <- runPCA(sce)
sce <- runTSNE(sce)
reducedDimNames(sce)
```

The provided cell annotations for this data set are available in `colData(sce)`. 

```{r}
colnames(colData(sce))
```

### Generating color maps

We define color palettes to use in `r Biocpkg("iSEE")` when coloring according to specific cell annotations or expression values. 
Each palette should be a function that accepts a number and returns a vector of colors of length equal to the input number.

For categorical variables, the function should accept the number of levels and return a color per level.
This is illustrated below with a function that generates colors for the `driver_1_s` factor.

```{r allen-dataset-3}
driver_color_fun <- function(n){
  return(RColorBrewer::brewer.pal(n, "Set2"))
}
```

Alternatively, functions can just return a named vector of colors if users want to specify the color for each level explicitly.
In such cases, the function can ignore its argument, though it is the user's responsibility to ensure that all levels are accounted for:

```{r}
qc_color_fun <- function(n){
  qc_colors <- c("forestgreen", "firebrick1")
  names(qc_colors) <- c("Y", "N")
  return(qc_colors)
}
```

For continuous variables, the function will be asked to generate a number of colors (21 by default).
Interpolation will then be performed internally to generate a color gradient.
Users can use existing color scales like `viridis::viridis`, or make their own interpolation points with a function that simply ignores the input number:

```{r}
fpkm_color_fun <- function(n){
  x <- c("black","brown","red","orange","yellow")
  return(x)
}
```

These values are stored in an object of the `ExperimentColorMap` class. 
Named functions passed to `assays` or `colData` will be used for coloring those specific assays or column data.

```{r}
ecm <- ExperimentColorMap(
  assays = list(
      counts = viridis::viridis,
      cufflinks_fpkm = fpkm_color_fun
  ),
  colData = list(
      passes_qc_checks_s = qc_color_fun,
      driver_1_s = driver_color_fun
  )
)
ecm
```

Users can change the defaults for assays or column data without explicit color mappings. 
Defaults for all variables can also be altered.
The example shown below alters the defaults for continuous variables, but the same applies for categorical factors as well. 

```{r}
ExperimentColorMap(
  all_continuous=list(
    assays=viridis::plasma,
    colData=viridis::inferno
  ),
  global_continuous=viridis::magma  
)
```

### Exploring the dataset

To begin the exploration, we create an `iSEE` app with the `SingleCellExperiment` object and the colormap generated above.

```{r allen-dataset-4}
app <- iSEE(sce, colormap = ecm)
```

We run this using `runApp` to open the app on our browser.

```{r, eval=FALSE}
shiny::runApp(app)
```

By default, the app starts with a dashboard containing one reduced dimension plot, one column data plot, one gene expression plot and one gene statistics table (see Use case II for how to change this). 
By opening the collapsible panels named "Plotting parameters", "Coloring parameters" and "Brushing parameters" under each plot, we can see and change what is being displayed in the respective plots, and also how it is being displayed. 

Let us start by exploring the reduced dimension plot. As can be seen from the "Plotting parameters" panel, this plot shows the first two principal components. Change `Type` to `(2) TSNE` to instead see the two-dimensional t-SNE representation. Next, open the "Coloring parameters" panel. By default, the points (cells) are not colored. By selecting `Column data` and choosing one of the variables in the dropdown menu that shows up, the cells can be colored by any of the provided annotations. Let’s choose `passes_qc_checks_s`, one of the annotations for which we defined a custom color map in the preparations above. Now, all cells that passed QC (`Y`) are colored forestgreen, while the ones that didn’t pass are colored firebrick. 

Now let’s move on to the column data plot. Here, we see the distribution of the number of reads (`NREADS`) across the cells in the data set, as well as the individual values for each cell. Note that the location of the points along the x-axis is generated by the jittering, and does not encode any information (you can see this since `X-axis = None` in the Plotting parameters panel). We can also plot two cell annotation against each other, by setting `X-axis` to be `Column data` and choosing one of the variables in the drop down menu that pops up. For example, we can choose `NALIGNED` (the number of aligned reads), and we see that (as expected), there is a very strong association between the total number of reads and the number of aligned reads. Again, we can color the cell by whether or not they passed the QC, by selecting `Column data` in the "Coloring parameters" panel and choosing `passes_qc_checks_s` in the dropdown menu. 

Finally, we will look at the gene expression plot. This plot displays the distribution of the expression values for a given gene, which has been specified by selecting the first row in the gene statistics table (this is seen in the "Plotting parameters" tab, since `Y-axis` is `Gene table`, and it is indicated that the `Y-axis gene linked to` is `Gene statistics table 1`. The values shown in the plots are taken from the `logcounts` assay of the provided `SingleCellExperiment`. We can modify this choice to any other assay that was available in the object (here, e.g., to `rsem_tpm`) to display other expression values. Note how the y-axis title keeps track of what is displayed in the plot. Again, note that all assays must be precalculated before the object is passed to `iSEE`. If we would like to show the expression of a particular gene of interest (say, `Znrf1`), we can either find and select it in the gene statistics table (use the search box just above the table), or we can set `Y-axis` to `Gene text`, and type `Znrf1` in the text box that shows up. As for the other plots, we can color in many ways. Let us color the points in the gene expression plot by the expression of another gene (say, `Grp`). To do this, open up the "Coloring parameters" panel, select `Gene text`, and paste `Grp` in the text box that shows up. 

After exploring the individual plots, we will now see how they can be linked together using the brushing functionality. To this end, let’s say that we are interested in seeing the expression of a certain gene in a particular cluster of cells from the reduced dimension plot. First, open the "Brushing parameters" panel under the Gene expression plot, and choose `Receive brush from:` to be `Reduced dimension plot 1` and set the `Brush effect` to `Transparent`. Then, drag the mouse to draw a rectangle around the cluster of interest in Reduced dimension plot 1. You will see that the points in `Gene expression plot 1` that are not within the rectangle in the `Reduced dimension plot 1` are now more transparent, allowing you to see the distribution of expression values for the chosen gene in the cluster of interest. By changing the `Brush effect`, you can also restrict the receiving plot to only show the selected points (`Restrict`) or color the selected points (be careful with the color choice if you have already colored the points according to another covariate via the "Coloring parameters" panel). 

## Use case II: Changing the default start configuration

The default start configuration with one plot of each type is not optimal for all use cases.
`iSEE` allows the user to programmatically modify the initial settings, avoiding the need to click through the choices to obtain the desired panel setup.

To demonstrate, let's say that we are only interested in gene expression plots.
The default set of panels can be changed via the `initialPanels` argument to `iSEE()`. 
Given a `SingleCellExperiment`/`SummarizedExperiment` `sce`, the following code opens an app with two adjacent gene expression plots. 

```{r}
init <- DataFrame(
  Name = c("Gene expression plot 1", "Gene expression plot 2"),
  Width = c(6, 6)
)
app <- iSEE(sce, initialPanels = init)
```

The genes to show on the y-axis in the two plots can be specified via the `geneExprArgs` argument to `iSEE`. 
This is a `DataFrame`, specifying the initial parameters for each plot. 
In this case, we want to modify the `YAxis` and `YAxisGeneText` defaults for the two plots:

```{r}
gexArg <- DataFrame(
  YAxis = "Gene text", 
  YAxisGeneText = c("0610009L18Rik", "0610009B22Rik")
)
app <- iSEE(sce, initialPanels = init, geneExprArgs = gexArg)
```

This will open the app with two gene expression plots, showing the selected genes. 
Of course, from this starting point, all the interactive functionality is still available, and new panels can be added, modified and linked via brushing. 
Users are referred to `?defaults` for the full list of values that can be specified in `iSEE`.

## Use case III: Exploring mass cytometry data

The flexibility of the `iSEE` interface means that it can be used to explore a variety of data types that can be represented in tabular format. 
Here we show how `iSEE` can be used to explore a mass cytometry (CyTOF) data set (originally from [@bodenmiller2012cytof]), by defining an initial panel configuration mimicking the familiar gating setup, consisting of multiple two-dimensional scatter plots.
The example data set was prepared as described by [@nowicka2017workflow], and the code to generate a `SummarizedExperiment` object is available from [https://github.com/lmweber/CyTOF-example-data](https://github.com/lmweber/CyTOF-example-data). 
Here we use a subset of the original data, consisting of 5,000 randomly selected cells, for which 35 markers are measured. 
The `SummarizedExperiment` object also contains a t-SNE representation of the 5,000 cells and the results of a cell clustering performed with `r Biocpkg("FlowSOM")` [@vangassen2015flowsom] in order to automatically detect cell populations. The data set can be downloaded as follows:

```{r, eval=FALSE}
cytof <- readRDS(gzcon(url("https://www.dropbox.com/s/42kfedfpzyssqvq/cytof_bcrxl_5000cells.rds?dl=1")))
```

The traditional way to analyze cytometry experiments and identify known cell populations is to construct a collection of two-dimensional scatter plots, each visualizing the cells in the space of two of the measured markers. 
By successively selecting subsets of cells with specific patterns of marker abundance ("gating"), experienced users can assign cells to known populations. 
To mimic part of the original gating strategy from [@bodenmiller2012cytof], we define a starting configuration for `iSEE` with five gene expression plots, each showing two of the markers. 
Each plot is linked to the previous one, with the brushing effect set to "Restrict", implying that only the cells selected by brushing in the previous plot will be shown.
We also include a reduced dimension plot, in which the cells are colored by the automatic cluster assignment. 
The Reduced dimension plot receives a restricting brush from the last Gene expression plot, which means that only the points that are selected by brushing in the latter will be shown in the Reduced dimension plot. 

The code below sets up the initial panel configuration and brushing relationships. 


```{r, eval=FALSE}
initialPanels = DataFrame(Name = c(paste("Gene expression plot", 1:5), 
                                   "Reduced dimension plot 1"),
                          Width = c(rep(3, 5), 5))
gexArg <- DataFrame(XAxis = "Gene text", YAxis = "Gene text",
                    XAxisGeneText = c("Cell_length", "CD3", "CD7", "CD3", "CD4"), 
                    YAxisGeneText = c("CD45", "CD20", "CD45", "CD33", "CD45"),
                    BrushByPlot = c("", paste("Gene expression plot", 1:4)),
                    BrushEffect = "Restrict")
redArg <- DataFrame(ColorBy = "Column data", 
                    ColorByColData = "population", 
                    BrushByPlot = "Gene expression plot 5",
                    BrushEffect = "Restrict")
```

```{r, eval=FALSE}
iSEE(cytof, initialPanels = initialPanels, geneExprArgs = gexArg, redDimArgs = redArg)
```

Once we have started `iSEE` in this configuration, we can start brushing from the first gene expression plot to successively refine a cell subpopulation. 
Roughly following the gating strategy in [@bodenmiller2012cytof], we select cells with high CD45 expression in `Gene expression plot 1`, followed by cells with low CD20 expression in `Gene expression plot 2`. 
Next, we select cells with high CD7 expression in `Gene expression plot 3`, and subsequently cells with high CD3 and low CD33 expression in `Gene expression plot 4`. 
Finally, by selecting either cells with high CD4 expression or those with low CD4 expression in `Gene expression plot 5`, and reviewing the resulting cells shown in the `Reduced dimension plot 1`, we can see that these are indeed the cells that were automatically classified as CD4+ (or CD8+) T-cells by the `r Biocpkg("FlowSOM")` clustering. 

At this point, we can export all the code needed to reproduce the six panels by clicking on the wrench in the top right corner and selecting "Extract the R code". 


## Use case III: Using the `ExperimentHub`

Here we will use `r Biocpkg("iSEE")` to showcase a TCGA RNA-seq data set of
`r Biocpkg("Rsubread")`-summarized raw count data for 7,706 tumor samples,
represented as an `ExpressionSet`.
The R data representation was derived from GEO accession GSE62944.
It is here retrieved from the Bioconductor `r Biocpkg("ExperimentHub")`,
immediately ready for post-processing and visualisation.

```{r ExperimentHub, eval=FALSE}
stopifnot(
  require(ExperimentHub),
  require(SingleCellExperiment),
  require(irlba),
  require(Rtsne),
  requireNamespace("scater")
)
ehub <- ExperimentHub()
eh1 <- ehub[["EH1"]] # an ExpressionSet
se1 <- as(eh1, "SummarizedExperiment")
sce1 <- as(se1, "SingleCellExperiment")
sce1 <- scater::runPCA(sce1, exprs_values = "exprs")
irlba_out <- irlba(exprs(sce1))
tsne_out <- Rtsne(irlba_out$v, pca = FALSE, perplexity = 50, verbose = TRUE)
reducedDim(sce1, "TSNE") <- tsne_out$Y
if (interactive()) { iSEE(sce1) }
```

## Using online file sharing systems

Active projects may require regular updates to the `SingleCellExperiment`
object, that must be iteratively re-distributed to collaborators.

One possible choice is to store the up-to-date `SingleCellExperiment` object
in an file produced by the `saveRDS` command, and to host this file in
an online file sharing system (*e.g.*, Drobpox).

Users may then download the file and launch the `iSEE` application as follows:

```{r curl_download, eval=FALSE}
library(curl)
rdsURL <- "https://my.file.sharing.website.com/URI"
temp_file <- tempfile(pattern = "iSEE_", fileext = ".rds")
message("Downloading URL to temporary location: ", temp_file)
curl_download(url = rdsURL, destfile = temp_file, quiet = FALSE)
sce <- readRDS(temp_file)
if (interactive()) { iSEE(sce) }
```

To save the downloaded file to a persistent location, users may adapt
the following code chunk:

```{r copy_curl, eval=FALSE}
newLocation <- "/path/of/your/choice/new_file_name.rds"
file.copy(from = temp_file, to = newLocation)
```

# FAQ

**Q: Can you implement a 'Copy to clipboard' button in the code editor?**

A: This is not necessary, as one can click anywhere in the code editor and
instantly select all the code using a keyboard shortcut that depends on your
operating system.

**Q: When brushing with a transparency effect, it seems that data points in the
receiving plot are not subsetted correctly.**

A: The subsetting is correct. What you see is an artefact of overplotting:
in areas excessively dense in points, transparency ceases to be an effective
visual effect.

**Q: Brushing on violin or square plots doesn't seem to select anything.**

A: For violin plots, points will be selected only if the brushed area includes
the center of the x-tick, i.e., the center of the violin plot.
This is intentional as it allows easy selection of all points in complex grouped
violin plots.
Indeed, the location of a specific point on the x-axis has no meaning.
The same logic applies to the square plots, where only the center of each square
needs to be selected to obtain all the points in the square.

-> to mention:
Subset and violins
Brushing backwards does not seem to work

# Additional information

Bug reports can be posted on the
[Bioconductor support site](https://support.bioconductor.org) or raised as
issues in the `r Githubpkg("csoneson/iSEE")` GitHub repository.
The GitHub repository also contains the development version of the package,
where new functionality will be added in the future.
The authors appreciate well-considered suggestions for improvements or new
features, or even better, pull requests.

If you use `r Biocpkg("iSEE")` for your analysis, please cite it as shown below:

```{r citation}
citation("iSEE")
```

# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```

# References {.unnumbered}

