---
title: "iSEE User's Guide"
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: aaron.lun@cruk.cam.ac.uk
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevin.rue-albrecht@kennedy.ox.ac.uk
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
abstract: >
  Abstract for iSEE
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{iSEE User's Guide}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---
<!-- as biocviews: -->
<!-- Visualization, GUI, DimensionReduction,  -->
<!--         FeatureExtraction, Clustering, Transcription,  -->
<!--         GeneExpression, Transcriptomics, SingleCell -->

<!-- **Package**: `r Rpackage("iSEE")` -->

<!-- **Authors**: `r packageDescription("iSEE")[["Author"]]` -->

<!-- **Version**: `r packageDescription("iSEE")$Version` -->

**Compiled date**: `r Sys.Date()`

**Last edited**: 2017-12-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Getting started

<!-- logo? MOS DEF :) -->

`r Biocpkg("iSEE")` is an R package distributed as part of the
[Bioconductor](http://bioconductor.org) project. To install the package,
start R and enter:

```{r installation, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
```

To load and attach the package to your current workspace, enter:

```{r library, message=FALSE, warning=FALSE}
library("iSEE")
```

# Introduction

`r Biocpkg("iSEE")`(Interactive SummarizedExperiment Explorer) is an R/Bioconductor package that provides an interactive Shiny-based graphical user interface for exploring data stored in SummarizedExperiment objects, including row- as well as column-level metadata. Particular attention is given to single-cell data in a `SingleCellExperiment` object with visualization of dimensionality reduction results (derived e.g. from Principal Components Analysis, or t-Stochastic Neighbour Embedding []vandermaaten[]). 
The generated plots can be conveniently linked to each other by brushing (i.e. clicking and holding the mouse over a plot area), and with a nice touch of reproducibility, you can obtain the full code to re-create the plots at the end of your live session. The combination of interactivity and reproducibility follows the proposal in the `pcaExplorer` and `ideal` packages, to deliver an immediate user experience, and guarantee the aspect of technical reproducibility of the generated output.

## Citation info

If you use `r Biocpkg("iSEE")` for your analysis, please cite it as here below:

```{r citation}
citation("iSEE")
```

# Quick start

Do you want to dive deep in using `r Biocpkg("iSEE")` right away? Here's what you need to do:

- install and load the `r Biocpkg("iSEE")` package
```{r qs-load, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
library("iSEE")
```

- get a dataset to work with, as for example the `allen` dataset from the `r Biocpkg("scRNAseq")` package, and perform the following operations to transform it into a `SingleCellExperiment` object, where you can store the reduced dimension representations (PCA and t-SNE)

```{r qs-data}
library(scRNAseq)
data(allen)
library(scater)
sce <- as(allen, "SingleCellExperiment")
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)
sce <- runPCA(sce)
sce <- runTSNE(sce)
sce
```

- launch the app itself

```{r qs-launch, eval=FALSE}
iSEE(sce)
```

- once you have started the app, look in the upper right corner for a question mark icon. Click on that, and be ready to perform an interactive tour of the app, based on the `rintrojs` package. []ganz[] During this tour, you will be showed around the different components of iSEE and learn the basic usage mechanisms by doing: the highlighted elements will be responding to the user's actions, while the rest of the UI will be shaded. You can move forward and backward along the tour by clicking on the `Next`/`Back` buttons, or also using the arrow keys. To exit the tour, either click on `Skip`, or simply click outside of the highlighted UI element.  

Once you are done generating the plots, do not forget to export the code that generated the outputs, by clicking on the sidebar button with the magic wand icon. This will open a modal popup where the code is displayed and automatically highlighted in a `ShinyAce`-based text editor. Select it all, or parts of it, copy it to the clipboard, and paste it into your analysis script/Rmarkdown file.


# Using the application

This section explains more in detail the different parameters you can adjust when launching the `iSEE` app. This set refers to a local instance of `iSEE`.

- `se` is a `SingleCellExperiment` object, the only mandatory parameter for launching `iSEE`

- `redDimArgs` is a `DataFrame` similar to that produced by the `redDimPlotDefaults` function, specifying initial parameters for the reduced dimension plots. Correspondingly, `colDataArgs` (e.g. created via `colDataPlotDefaults`) specifies the initial parameters for the column data plots, and `geneExprArgs` (via `geneExprPlotDefaults`) for the gene expression plots. `geneStatArgs` (a `DataFrame` structured as the output of  `geneStatTableDefaults`) specifies the initial parameters for the gene statistics tables
- `redDimMax`, `colDataMax`, and `geneExprMax` are integer scalars specifying the maximum number of plots (respectively, reduced dimension, column data, and gene expression) in the main interface. `geneStatMax`, similarly, is an integer scalar specifying the maximum number of gene statistic tables in the interface
- `initialPanels` is a `DataFrame` specifying which panels should be created at the  initialization. This should contain a `Name` character field (specifying the identity of the panel, e.g., “Reduced dimension plot 1”, “Gene statistics table 2”) and a `Width` integer field (ranging from 2 to 12). If not specified, it will generate one plot per type, together with one gene stats table  
- `annot.orgdb` is an `org.*.eg.db` annotation object from which Entrez identifiers can be retrieved. If this is specified, gene information will be retrieved upon selection of particular genes in the data table.
- `annot.keytype` is a string specifying the keytype to use to query `annot.orgdb`, while `annot.keyfield` is a string specifying the field of `rowData(se)` containing the keys of type `annot.keytype`. If `NULL`, the row names of `se` are used as the keys.
- `colormap` is an `ExperimentColorMap` object that defines custom color maps to apply to individual `assays`, `colData`, and `rowData` covariates. For more details on how to create this object, please consult the example of `iSEE`

Upon calling `iSEE`, a Shiny App is launched for interactive data exploration of the `SingleCellExperiment`/`SummarizedExperiment` object.


## Using `iSEE` from a demo public instance

[]if we want to allow this[]

## Deploying `iSEE` to your Shiny Server

[]same here[]



# Getting to know the user interface and the functionality

Probably the most immediate way to master the user interface of `iSEE` is by taking the guided interactive tour, as this feature provides a pretty exhaustive showcase of the relevant elements and possible user interactions. Nonetheless, a brief description of the UI is provided in this section.

header, sidebar, main panel

## Specifical subsection on the help/tour?


# Functionality

aka description of plot types

concept of brushing and linking - here?

this section has for each type of plot a screenshot of plot + plot settings


# Bug reporting/Devel Version


The GitHub repository for `r Biocpkg("iSEE")` is https://github.com/csoneson/iSEE, where you can find the development version of the package. This is also the place to file an issue, report a bug, or provide a pull request.


# Use cases

## Allen dataset





## TCGA dataset


-> link to youtube screencast where we do it?!



## Using the `ExperimentHub`

```{r ExperimentHub, eval=FALSE}
stopifnot(
  require(ExperimentHub),
  require(SingleCellExperiment),
  require(irlba),
  require(Rtsne),
  requireNamespace("scater")
)
ehub <- ExperimentHub()
eh1 <- ehub[["EH1"]] # an ExpressionSet
se1 <- as(eh1, "SummarizedExperiment")
sce1 <- as(se1, "SingleCellExperiment")
sce1 <- scater::runPCA(sce1, exprs_values = "exprs")
irlba_out <- irlba(sce1@assays@.xData$data$exprs)
tsne_out <- Rtsne(irlba_out$v, pca = FALSE, perplexity = 50, verbose = TRUE)
reducedDim(sce1, "TSNE") <- tsne_out$Y
if (interactive()) { iSEE(sce1) }
```


## Using online file sharing systems

Active projects may require regular updates to the `SingleCellExperiment`
object, that must be iteratively re-distributed to collaborators.

One possible choice is to store the up-to-date `SingleCellExperiment` object
in an file produced by the `saveRDS` command, and to host this file in
an online file sharing system (*e.g.*, Drobpox).

Users may then download the file and launch the `iSEE` application as follows:

```{r curl_download, eval=FALSE}
library(curl)
rdsURL <- "https://my.file.sharing.website.com/URI"
temp_file <- tempfile(pattern = "iSEE_", fileext = ".rds")
message("Downloading URL to temporary location: ", temp_file)
curl_download(url = rdsURL, destfile = temp_file, quiet = FALSE)
sce <- readRDS(temp_file)
if (interactive()) { iSEE(sce) }
```

To save the downloaded file to a persistent location, users may adapt
the following code chunk:

```{r copy_curl, eval=FALSE}
newLocation <- "/path/of/your/choice/new_file_name.rds"
file.copy(from = temp_file, to = newLocation)
```



# Further development

Additional functionality for the `r Biocpkg("iSEE")` will be added in the future.

Improvements, suggestions, bugs, issues and feedback of any type can be sent to marinif@uni-mainz.de.


# FAQ

**Q: Can you implement a 'Copy to clipboard' button in the code editor?**

A: This is not necessary, as one can click anywhere in the code editor and
instantly select all the code using a keyboard shortcut that depends on
your operating system.

**Q: When brushing with a transparency effect,
it seems that data points in the receiving plot are not subsetted correctly.**

A: The subsetting is correct. What you see is an artefact of overplotting:
in areas excessively dense in points, transparency ceases to be an effective
visual effect.





# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```


# References {.unnumbered}
 -> if we use them








