---
title: "iSEE User's Guide"
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: aaron.lun@cruk.cam.ac.uk
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevin.rue-albrecht@kennedy.ox.ac.uk
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
abstract: >
  Abstract for iSEE
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{iSEE User's Guide}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: iSEE.bib
---
<!-- as biocviews: -->
<!-- Visualization, GUI, DimensionReduction,  -->
<!--         FeatureExtraction, Clustering, Transcription,  -->
<!--         GeneExpression, Transcriptomics, SingleCell -->

<!-- **Package**: `r Rpackage("iSEE")` -->

<!-- **Authors**: `r packageDescription("iSEE")[["Author"]]` -->

<!-- **Version**: `r packageDescription("iSEE")$Version` -->

**Compiled date**: `r Sys.Date()`

**Last edited**: 2017-12-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Getting started

<!-- logo? MOS DEF :) -->

`r Biocpkg("iSEE")` is an R package distributed as part of the
[Bioconductor](http://bioconductor.org) project. To install the package,
start R and enter:

```{r installation, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
```

To load and attach the package to your current workspace, enter:

```{r library, message=FALSE, warning=FALSE}
library("iSEE")
```

# Introduction

`r Biocpkg("iSEE")`(Interactive SummarizedExperiment Explorer) is an R/Bioconductor package that provides an interactive Shiny-based graphical user interface for exploring data stored in SummarizedExperiment objects, including row- as well as column-level metadata. Particular attention is given to single-cell data in a `SingleCellExperiment` object with visualization of dimensionality reduction results (derived e.g. from Principal Components Analysis, or t-distributed Stochastic Neighbour Embedding []vandermaaten[]). 
The generated plots can be conveniently linked to each other by brushing (i.e. clicking and holding the mouse over a plot area), and with a nice touch of reproducibility, you can obtain the full code to re-create the plots at the end of your live session. The combination of interactivity and reproducibility follows the proposal in the `pcaExplorer` and `ideal` packages, to deliver an immediate user experience, and guarantee the aspect of technical reproducibility of the generated output.

## Citation info

If you use `r Biocpkg("iSEE")` for your analysis, please cite it as here below:

```{r citation}
citation("iSEE")
```

# Quick start {#quickstart}

Do you want to dive deep in using `r Biocpkg("iSEE")` right away? Here's what you need to do:

- install and load the `r Biocpkg("iSEE")` package
```{r qs-load, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("iSEE")
library("iSEE")
```

- get a dataset to work with, for example the `allen` dataset from the `r Biocpkg("scRNAseq")` package, and perform the following operations to transform it into a `SingleCellExperiment` object, where you can store the reduced dimension representations (here, PCA and t-SNE)

```{r qs-data, message=FALSE}
library(scRNAseq)
data(allen)
library(scater)
sce <- as(allen, "SingleCellExperiment")
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)
sce <- runPCA(sce)
sce <- runTSNE(sce)
sce
```

- launch the app itself

```{r qs-launch, eval=FALSE}
iSEE(sce)
```

- once you have started the app, look in the upper right corner for a question mark icon. Click on that, and be ready to perform an interactive tour of the app, based on the `rintrojs` package. []ganz[] During this tour, you will be taken through the different components of iSEE and learn the basic usage mechanisms by doing: the highlighted elements will be responding to the user's actions, while the rest of the UI will be shaded. You can move forward and backward along the tour by clicking on the `Next`/`Back` buttons, or also using the arrow keys. To exit the tour, either click on `Skip`, or simply click outside of the highlighted UI element.  

Once you are done generating the plots, you can export the code that generated the outputs, by clicking on the sidebar button with the magic wand icon. This will open a modal popup where the code is displayed and automatically highlighted in a `ShinyAce`-based text editor. Select it all, or parts of it, copy it to the clipboard, and paste it into your analysis script/Rmarkdown file.


# Using the application

This section explains more in detail the different parameters you can adjust when launching the `iSEE` app. This set refers to a local instance of `iSEE`.

- `se` is a `SingleCellExperiment` object, the only mandatory parameter for launching `iSEE`

- `redDimArgs` is a `DataFrame` similar to that produced by the `redDimPlotDefaults` function, specifying initial parameters for the reduced dimension plots. Correspondingly, `colDataArgs` (e.g. created via `colDataPlotDefaults`) specifies the initial parameters for the column data plots, and `geneExprArgs` (via `geneExprPlotDefaults`) for the gene expression plots. `geneStatArgs` (a `DataFrame` structured as the output of  `geneStatTableDefaults`) specifies the initial parameters for the gene statistics tables
- `redDimMax`, `colDataMax`, and `geneExprMax` are integer scalars specifying the maximum number of plots (respectively, reduced dimension, column data, and gene expression) in the main interface. `geneStatMax`, similarly, is an integer scalar specifying the maximum number of gene statistic tables in the interface
- `initialPanels` is a `DataFrame` specifying which panels should be created at the  initialization. This should contain a `Name` character field (specifying the identity of the panel, e.g., “Reduced dimension plot 1”, “Gene statistics table 2”) and a `Width` integer field (ranging from 2 to 12). If not specified, it will generate one plot per type, together with one gene stats table  
- `annot.orgdb` is an `org.*.eg.db` annotation object from which Entrez identifiers can be retrieved. If this is specified, gene information will be retrieved upon selection of particular genes in the data table.
- `annot.keytype` is a string specifying the keytype to use to query `annot.orgdb`, while `annot.keyfield` is a string specifying the field of `rowData(se)` containing the keys of type `annot.keytype`. If `NULL`, the row names of `se` are used as the keys.
- `colormap` is an `ExperimentColorMap` object that defines custom color maps to apply to individual `assays`, `colData`, and `rowData` covariates. For more details on how to create this object, please consult the example of `iSEE`

Upon calling `iSEE`, a Shiny App is launched for interactive data exploration of the `SingleCellExperiment`/`SummarizedExperiment` object.


## Using `iSEE` from a demo public instance

[]if we want to allow this[]

## Deploying `iSEE` to your Shiny Server

[]same here[]



# Getting to know the user interface and the functionality

Probably the most immediate way to master the user interface of `iSEE` is by taking the guided interactive tour, as this feature provides a pretty exhaustive showcase of the relevant elements and possible user interactions. Nonetheless, a brief description of the UI (based on the `shinydashboard` package) is provided in this section.

## The `iSEE` header

The dashboard header for `iSEE` contains, apart from the title, a dropdown menu, where a number of help elements are available:

- the “Click me for a quick tour” button activates the `rintrojs` based tour. Some more detail on this is included in Section \@ref(quickstart)
- The “Open the vignette (web)” button opens in a new tab the vignette for `iSEE` from the Bioconductor website (internet connection is required)
- The “Open the vignette (local)” button is a fallback solution that fetches the HTML vignette of `iSEE` as it is available on the system
- “About iSEE” displays a modal popup containing information about how to cite `iSEE`, as well as the output from `sessionInfo()`, which should be included in case of reporting a bug/issue


## The `iSEE` sidebar

In the dashboard sidebar of `iSEE`, 4 buttons allow the creation of new plot elements in the main body of the app. The maximum number of plots per type can be defined when launching the app. Upon clicking either of these buttons, additional plots or tables are inserted, together with placeholders to reorganize the sequence of how they should appear. This can be controlled by means of the up and down arrow buttons, and unnecessary plots can be removed by clicking on the trash button. Also, the width of each panel can be adjusted, in a value ranging from 2 to 12 (12 being the width of a full `fluidRow`).

To help the user in organizing the plots, a color coding reminds the plot type being moved/canceled []TODO[]

A fundamental element in the sidebar is the `Extract the R code!` button (with the magic wand icon). By clicking on this, the user opens up a modal popup window, where an instance of a text editor (based on the `ShinyAce` package) contains the code snippet required to reproduce the entire plots, created in the live session. This can be typically used at the end of each exploration, to extract a reproducible snapshot of the commands, which can be furthermore edited, in order to finalize the plots (e.g. to get ready for publication). 



## The `iSEE` dashboard body

The main element in the body of `iSEE` is the combination of plots and tables, generated (and optionally linked to one another) according to the different actions taken by the user. The main explanation on how the different plots and tables work is presented in Section \@ref(functionality).

At the bottom of the body, a footer including some information on the `iSEE` development team can be found.



# Functionality {#functionality}

This section contains a description of each plot and table type you can generate with `iSEE`. For each plot panel, three different sets of parameters will be available in collapsible boxes, namely `Plotting parameters`, `Coloring parameters`, and `Brushing parameters`. While the plotting parameters will be specific to each plot type, the coloring and brushing panels look the same for all types. How the brushing operates is described in additional detail in Section \@ref(brushing).

## Reduced dimension plots

[]screenshot plot + plot params[]
You can obtain a reduced dimension plot (e.g. PCA, t-SNE) of your expression dataset to efficiently explore the high-dimensional dataset under inspection. The plotting parameters control the `redDim` slot to be displayed, as well as the two dimensions to plot against each other. 

You can color your points (i.e. cells or samples) in different ways. Default is no color scheme (`None` in the radio button), but it is possible to use the value of any `Column data`, be it discrete or continuous. The plotting routine recognizes this, and adapts automatically the scale to use. Another possibility is to plot by expression value of a particular gene, either via the linked `Gene table`, or via the `Gene text` which is input in the text field. The selection via gene table is made possible by clicking on one row in the corresponding `datatable` element. Among the different choices for expression values, it is possible to select one of the `assays` in the provided `SingleCellExperiment`/`SummarizedExperiment`.  


## Column data plots

[]screenshot plot + plot params[]

You can also have a plot where the variable of interest (on the y-axis) is one of the experimental covariates (here called a column data plot). The x-axis can be chosen as `None` or also specify another experimental covariate. 

According to the combination of variable types, the plot can assume different forms: a violin plot for a continuous covariate, grouped violins if this is split by some categorical factor, or also a scatter plot if both are continuous. A categorical variable is represented as squared-box plots, which can also extend to two dimensions in case of two categorical variables. In the categorical VS continuous case, again we see grouped violin plots (in every case, superimposed jittered points are also present).

Again, the coloring options can be done by no grouping, column data (continuous and categorical), as well as expression values.


## Gene expression plots

[]screenshot plot + plot params[]

Last but not least, a gene expression plot represents the expression values for the samples at hand in different stratifications (if selecting an experimental factor for the x-axis) or as a scatter plot (if you chose another continuous variable, e.g. the expression values of another gene). 

The gene selection for the y-axis is done as in the other panels, i.e. either via clicking in the linked gene statistics table, or by entering the gene name as text. The measurement units are selected as one of the `assays(se)`, and the x-axis covariate follows the None/column data/expression value schema, also present in the other plots.

As in the other two plot types, the coloring options can be done by no grouping, column data (continuous and categorical), and expression values (linked to table or entered as text).


## Gene statistics tables

[]screenshot table?[]

A gene statistics table can contain by default the values of the `rowData` slot for the `SingleCellExperiment`/`SummarizedExperiment` object. If none are specified, a column named `Present` is added and set to `TRUE` for all available genes, in order to guarantee the functionality of a `datatable`.

Typically, these table can be used to handily connect to other plots, but also to retrieve on the fly (via the `rentrez` package, as well as linking to the NCBI database) some information on the selected gene, provided the `annot.orgdb`, `annot.keytype`, and `annot.keyfield` parameters are specified.


## Brushing plots: linking one plot to another {#brushing}

To link one plot to another, you need to perform two operations: activate the brush transmission from the source plot, and enable the brush receiving in the destination plot. This can be done respectively by clicking on the `Transmit brush` checkbox (source plot), and selecting the corresponding brush element in the `Receive brush from` selection widget (destination plot).

The two operations will allow you to link the two plots together. To actually do that, go on the source plot, and select an area in this plot which contains a subset of cells. If you correctly linked the two plots together, you should see that the destination plot is affected by the brushed selection.

The set of cells/samples you selected in the source plot will be affected in the destination plot, in the desired manner:

- `Restrict` will only plot the selected subset of points in the destination plot
- `Color` will mark the set of points in the destination plot with a color that can be selected via the `colourpicker` package 
- `Transparent` will draw the selected points with no transparency, while all non selected points will be plotted with an alpha value that can be specified in the slider

As a bonus, interacting with the plot as zooming in and out is possible by first brushing then double-clicking on the brushed area (when the `Zoom upon brush` checkbox is selected). To zoom out to the original plot, double-click on the plot. 

Warning: when brushing violin plots, points will be selected only if the brushed area includes the center of the x-tick a.k.a. the center of the violin plot - the current implementation is robustly defined with this way of selecting points.



# Bug reporting/Devel Version

The GitHub repository for `r Biocpkg("iSEE")` is https://github.com/csoneson/iSEE, where you can find the development version of the package. This is also the place to file an issue, report a bug, or provide a pull request.


# Use cases

## Allen dataset


```{r allen-dataset,eval=FALSE}
library(scRNAseq)
data(allen)
class(allen)
library(scater)
sce <- as(allen, "SingleCellExperiment")
counts(sce) <- assay(sce, "tophat_counts")
sce <- normalize(sce)
sce <- runPCA(sce)
sce <- runTSNE(sce)
sce
qc_colors <- c("forestgreen", "firebrick1")
names(qc_colors) <- c("Y", "N")
driver_colors <- RColorBrewer::brewer.pal(3, "Set2")
names(driver_colors) <- unique(sce$driver_1_s)
ecm <- new("ExperimentColorMap",
   assays = list(
       counts = viridis::viridis(10),
       cufflinks_fpkm = c("black","brown","red","orange","yellow")
   ),
   colData = list(
       passes_qc_checks_s = qc_colors,
       driver_1_s = driver_colors
   )
# launch the app itself
if (interactive()) { iSEE(sce, colormap = ecm) }
```






-> link to youtube screencast where we do it?!

## Using the `ExperimentHub`

Here we will use `iSEE` showcasing TCGA dataset

```{r ExperimentHub, eval=FALSE}
stopifnot(
  require(ExperimentHub),
  require(SingleCellExperiment),
  require(irlba),
  require(Rtsne),
  requireNamespace("scater")
)
ehub <- ExperimentHub()
eh1 <- ehub[["EH1"]] # an ExpressionSet
se1 <- as(eh1, "SummarizedExperiment")
sce1 <- as(se1, "SingleCellExperiment")
sce1 <- scater::runPCA(sce1, exprs_values = "exprs")
irlba_out <- irlba(sce1@assays@.xData$data$exprs)
tsne_out <- Rtsne(irlba_out$v, pca = FALSE, perplexity = 50, verbose = TRUE)
reducedDim(sce1, "TSNE") <- tsne_out$Y
if (interactive()) { iSEE(sce1) }
```


## Using online file sharing systems

Active projects may require regular updates to the `SingleCellExperiment`
object, that must be iteratively re-distributed to collaborators.

One possible choice is to store the up-to-date `SingleCellExperiment` object
in an file produced by the `saveRDS` command, and to host this file in
an online file sharing system (*e.g.*, Drobpox).

Users may then download the file and launch the `iSEE` application as follows:

```{r curl_download, eval=FALSE}
library(curl)
rdsURL <- "https://my.file.sharing.website.com/URI"
temp_file <- tempfile(pattern = "iSEE_", fileext = ".rds")
message("Downloading URL to temporary location: ", temp_file)
curl_download(url = rdsURL, destfile = temp_file, quiet = FALSE)
sce <- readRDS(temp_file)
if (interactive()) { iSEE(sce) }
```

To save the downloaded file to a persistent location, users may adapt
the following code chunk:

```{r copy_curl, eval=FALSE}
newLocation <- "/path/of/your/choice/new_file_name.rds"
file.copy(from = temp_file, to = newLocation)
```



# Further development

Additional functionality for the `r Biocpkg("iSEE")` will be added in the future.

Improvements, suggestions, bugs, issues and feedback of any type can be sent to marinif@uni-mainz.de|whatever other emails|or no email but rather open an issue.


# FAQ

**Q: Can you implement a 'Copy to clipboard' button in the code editor?**

A: This is not necessary, as one can click anywhere in the code editor and
instantly select all the code using a keyboard shortcut that depends on
your operating system.

**Q: When brushing with a transparency effect,
it seems that data points in the receiving plot are not subsetted correctly.**

A: The subsetting is correct. What you see is an artefact of overplotting:
in areas excessively dense in points, transparency ceases to be an effective
visual effect.


-> to mention:
Brushing violins
Subset and violins
Brushing backwards does not seem to work



# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```


# References {.unnumbered}
 -> if we use them








